add Files, match, result exercise (#3)
ferrous-systems/rust-exercises: exercise-book/src/files-match-result.md

    handling.

-   when to use the `.unwrap()` method.

-   propagate an error with the `?` operator

-   return the `Option`-type.

-   do some elementary file processing (opening, reading to buffer,
    counting, reading line by line).

-   navigate the Rust `stdlib` documentation

-   add external dependencies to your project

## Task

Find the exercise template here [`../../exercise-templates/files-match-result`](../../exercise-templates/files-match-result)

Find the solution to the exercise here [`../../exercise-solutions/files-match-result`](../../exercise-solutions/files-match-result). You can run them with the following command:
`cargo run --example step_x`, where x is the number of the step. 

1. Fix the runtime error in the template code by correcting the file path. Handle the `Result` type that is returned from the `File::open()` with a match statement, so that the `.unwrap()` can be deleted. 

2. Read the content of the file to a buffer using [Read::read\_to\_string](https://doc.rust-lang.org/std/io/trait.Read.html#method.read_to_string). Propagate the Error with `?` to `fn main()`. Start with the code of Step 1. 
  

> 3. Use the [`lines()`](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.lines)-method to read the lines into a [BufReader](https://doc.rust-lang.org/std/io/struct.BufReader.html) and count them! Start with the code of Step 1.

Reviewer: ```suggestion
3. Use the [`lines()`](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.lines)-method to read the file line-by-line from a [BufReader](https://doc.rust-lang.org/std/io/struct.BufReader.html) and count how many lines there are. Start with the code of Step 1.
```

> 3. Take the code from Step 1 again, but now use the [`lines()`](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.lines)-method to instead read the file line-by-line from a [BufReader](https://doc.rust-lang.org/std/io/struct.BufReader.html). Use this to count how many lines there are.

Reviewer: ```suggestion
3. Take the code from Step 1 again, but now construct a [BufReader](https://doc.rust-lang.org/std/io/struct.BufReader.html) and use its [`lines()`](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.lines)-method to read the file line-by-line. Use this to count how many lines there are.
```

> 4. Filter out empty lines using [is\_empty](https://doc.rust-lang.org/std/string/struct.String.html#method.is_empty) and print the non-empty ones. Start with the code of Step 3.

Reviewer: ```suggestion
4. Based on Step 3, filter out empty lines using [is\_empty](https://doc.rust-lang.org/std/string/struct.String.html#method.is_empty) and print the non-empty ones.
```

> 5. Write a function that parses each line and returns `Some(url)` if the line is a URL, and `None` if it is not. Use the Url crate's [UrlType](https://docs.rs/url/2.1.1/url/). Start with the code of Step 4.

Reviewer: ```suggestion
5. Take your code from Step 4 and now write a function that parses each line and returns `Some(url)` if the line is a URL, and `None` if it is not. Use the Url crate's [UrlType](https://docs.rs/url/2.1.1/url/).

variants, and depending on what those variants are, the program may
continue in a different way.

The Option type can have the variant `Some(T)` or `None`. `T` is a type parameter that means some type should go here, we'll decide which one later.. It is used, when you have to handle optional values. For example
if you want to be able to leave a field of a struct empty, you assign the
`Option` type to it. If the field has a value, it is `Some(<value>)`, if
it is empty, it is `None`.

The variants of the `Result` type are `Ok(t)` and `Err(e)`. It is used to
handle errors. If an operation was successful, `Ok(t)` is returned. In
`Ok(t)`, `t` can be the empty tuple or a return value. In `Err(e)`, `e`
contains an error message that can be printed.

Both types can be used with the `match` keyword. The received value is
matched on patterns, each leads to the execution of a different
expression.

### How to use `match`

`match` is a way of control flow based on pattern matching. A pattern on
the left results in the expression on the right side.

 ```rust
let value = true;

match value {
    true => println!("This is true!"),
    false => println!("This is false!"),
}
```

> Other then with if/else, every case has to be handled explicitly, at

Reviewer: ```suggestion
Unlike with if/else, every case has to be handled explicitly, at
```

✅ Use `println!` to print the content of the `String` buffer.

<details>
  <summary>Click me</summary>

```rust, ignore
{{#include ../../exercise-solutions/files-match-result/examples/step_2.rs}}
```
</details>

### Task 3: Read the lines into a `BufReader` and count them!

✅ Add the following imports:

```rust
use std::io::{ BufReader, BufRead,};
```

✅ Take a look at the documentation of [BufReader](https://doc.rust-lang.org/std/io/struct.BufReader.html). BufReader is a struct that adds buffering to any reader. It implements the
[`BufRead`](https://doc.rust-lang.org/std/io/trait.BufRead.html#) trait. In short this means, that methods that are defined for `BufRead` can be used for `BufReader`. For example the [`lines()`](https://doc.rust-lang.org/std/io/trait.BufRead.html#method.lines) method.

✅ Construct a `BufReader` around the file.

✅ The `lines()`- method returns an Iterator over the file’s lines. Iterate over the lines with a for loop to count them.

✅ Print the number of lines the file contains.

> ✅ You don’t have to handle the `Result` that is returned from `.lines()`, why?

Reviewer: `.lines()` doesn't return a `Result`?

Author: I deleted the question as that is kept better for the discussion afterward

Reviewer: Ah, the `Item` you get out of the Iterator is a `Result`. Maybe that's what was meant here. Maybe...

```suggestion
✅ When you go through the for loop on `.lines()`, each time around the loop you get  a `std::io::Result<&str>` and not a string-slice (`&str`). Why? Where does the `BufReader` get its input from?
```

Reviewer: OK!
